packages/app/src/cli/services/bulk-operations/stage-file.ts

```diff
@@ -0,0 +1,176 @@
+  const target = validateStagedUploadResponse(response)
+
+  const fileContents = await prepareFileContents(buffer, filePath)
+  await uploadFileToStagedUrl(fileContents, target.url, target.parameters, filename)
+
+  return extractStagedUploadKey(target.parameters)
+}
+
+/**
+ * Converts JSON variables string to JSONL buffer format
+ */
+function convertJsonToJsonlBuffer(jsonVariables?: string): Buffer | undefined {
+  if (jsonVariables === undefined) {
+    return undefined
+  }
+
+  const variablesArray: {[key: string]: unknown}[] = JSON.parse(jsonVariables)
+  const jsonlContent = `${variablesArray.map((vars) => JSON.stringify(vars)).join('\n')}\n`
+  return Buffer.from(jsonlContent, 'utf-8')
+}
+
+/**
+ * Calculates the file size from buffer or file path
+ */
+async function calculateFileSize(buffer?: Buffer, filePath?: string): Promise<number> {
+  if (buffer) {
+    return buffer.length
+  }
+  if (filePath) {
+    return fileSize(filePath)
+  }
+  return 0
+}
+
+/**
+ * Requests a staged upload URL from Shopify
+ */
+async function requestStagedUpload(
+  adminSession: AdminSession,
+  filename: string,
+  size: number,
+): Promise<StagedUploadsCreateMutation> {
+  return adminRequestDoc<StagedUploadsCreateMutation, StagedUploadsCreateMutationVariables>({
+    query: StagedUploadsCreate,
+    session: adminSession,
+    variables: {
+      input: [
+        {
+          filename,
+          fileSize: size.toString(),
+          httpMethod: 'POST',
+          mimeType: 'text/jsonl',
+          resource: 'BULK_MUTATION_VARIABLES',
+        },
+      ],
+    },
+  })
+}
+
+/**
+ * Validates the staged upload response and extracts the target
+ */
+function validateStagedUploadResponse(response: StagedUploadsCreateMutation): {
+  url: string
+  resourceUrl: string
+  parameters: {name: string; value: string}[]
+} {
+  if (!response.stagedUploadsCreate) {
+    throw new AbortError('No response received from stagedUploadsCreate mutation')
+  }
+
+  if (response.stagedUploadsCreate.userErrors.length > 0) {
+    const errors = response.stagedUploadsCreate.userErrors
+      .map((error: {field?: string[] | null; message: string}) => error.message)
+      .join(', ')
+    throw new AbortError(`Failed to create staged upload: ${errors}`)
+  }
+
+  const target = response.stagedUploadsCreate.stagedTargets?.[0]
+  if (!target) {
+    throw new AbortError('No staged upload target returned from Shopify')
+  }
+
+  if (!target.url || !target.resourceUrl) {
+    throw new AbortError('Invalid staged upload target: missing required URLs')
+  }
+
+  // Type assertion: we've validated that url and resourceUrl are non-null strings
+  return {
+    url: target.url,
+    resourceUrl: target.resourceUrl,
+    parameters: target.parameters,
+  }
+}
+
+/**
+ * Prepares file contents for upload by reading from buffer or file path
+ */
+async function prepareFileContents(buffer?: Buffer, filePath?: string): Promise<Buffer> {
```

Comment:
Same thing here -- feels like this is more complex than it has to be for the first pass.